RavenDB-16446 Avoid using index temp files

ayende opened the pull request.

ayende committed:
RavenDB-16446 Avoid using index temp files

ppekrol commented on src/Raven.Server/Indexing/TempFileCache.cs:
```diff
@@ -0,0 +1,126 @@
+            string path = _options.TempPath.FullPath;
+            if (Directory.Exists(path) == false)
+                Directory.CreateDirectory(path);
+            foreach (string file in Directory.GetDirectories(path,  "*" + StorageEnvironmentOptions.DirectoryStorageEnvironmentOptions.TempFileExtension))
```
This is the same directory that Voron is using. Can we have a conflict with some file that voron is using?
Also... Voron clears that directory on startup, will we have any files here?

ayende replied:
That is why I'm using the temp extension.
I'm not _expecting_ to have any files here, but this is there to handle the case of us leaving files around because we crashes, so we'll clean them up.

ppekrol replied:
Ok, but can we have a conflict with some of the files that Voron is using? I do not recall, but double-checking.

ppekrol commented on src/Raven.Server/Indexing/TempFileCache.cs:
```diff
@@ -0,0 +1,126 @@
+        private readonly Queue<FileStream> _files = new Queue<FileStream>();
+        private readonly Queue<MemoryStream> _ms = new Queue<MemoryStream>();
+
+        private const long MaxFileSizeToReduce = 16 * Constants.Size.Megabyte;
```
MaxFileSizeToKeepInBytes

ppekrol commented on src/Raven.Server/Indexing/TempFileCache.cs:
```diff
@@ -0,0 +1,126 @@
+                }
+                else
+                {
+                    _files.Enqueue(new FileStream(file, FileMode.OpenOrCreate));
```
FileMode.Open?

ayende replied:
Not critical either way

ppekrol commented on src/Raven.Server/Indexing/TempFileCache.cs:
```diff
@@ -0,0 +1,126 @@
+            else
+            {
+                stream = SafeFileStream.Create(
+                    Path.Combine(_options.TempPath.FullPath, Guid.NewGuid() +  StorageEnvironmentOptions.DirectoryStorageEnvironmentOptions.TempFileExtension),
```
Can it be `Lucene_{GUID}.tmp` ?

ayende replied:
Yes

ppekrol commented on src/Raven.Server/Indexing/TempFileCache.cs:
```diff
@@ -0,0 +1,126 @@
+                _files.Count >= MaxFilesInCache)
+            {
+                var fileName = s.Name;
+                s.Dispose();
```
We had `DisposeFile` method. Probably better to have sth like that here as well.

ppekrol commented on src/Raven.Server/Indexing/TempFileCache.cs:
```diff
@@ -0,0 +1,126 @@
+                string fileName = null;
+                try
+                {
+                    fileName = file.Name;
```
We had `DisposeFile` method. Probably better to have sth like that here as well.

ppekrol commented on src/Raven.Server/Indexing/TempFileCache.cs:
```diff
@@ -0,0 +1,126 @@
+        private readonly Queue<MemoryStream> _ms = new Queue<MemoryStream>();
+
+        private const long MaxFileSizeToReduce = 16 * Constants.Size.Megabyte;
+        private const int MaxFilesInCache = 32;
```
MaxFilesToKeepInCache

ppekrol commented on src/Raven.Server/Indexing/VoronIndexOutput.cs:
```diff
@@ -18,60 +22,51 @@ public class VoronIndexOutput : BufferedIndexOutput
+                    // too big, copy the buffer to the file
+                    _file = _fileCache.RentFileStream();
+                    _ms.CopyTo(_file);
+                    _fileCache.ReturnMemoryStream(_ms);
```
_ms should be set to `null` afterwards?

ayende replied:
Yes!

ppekrol commented on src/Raven.Server/Indexing/VoronIndexOutput.cs:
```diff
@@ -85,21 +80,27 @@ public override void Seek(long pos)
             {
-                _file.SetLength(length);
+                if(_ms != null)
+                    _ms.SetLength(length);
```
what if that length exceeds the capacity?

ayende replied:
Cannot happen, this is used only to trim the file, not to extend it.
At any rate, that would increased the MemoryStream size.

ppekrol commented on src/Raven.Server/Indexing/LuceneVoronDirectory.cs:
```diff
@@ -14,6 +14,7 @@ public sealed unsafe class LuceneVoronDirectory : Lucene.Net.Store.Directory
         private readonly StorageEnvironment _environment;
         private readonly string _name;
         private readonly IndexOutputFilesSummary _indexOutputFilesSummary;
+        private readonly TempFileCache _tmpFileCache;
```
The problem that I can see here is that size of this can be 32 * 128kb = 4mb

Now if we will take that 4MB and multiply that by the amount of indexes then we are getting quite a lot of memory. Are we OK with that?
Do we want to subscribe to low memory notifications and clear the cache then?

ayende replied:
Good point, I'll add handling to that.

ayende committed:
RavenDB-16446 PR comments, adding low memory handling, fixing slow test compilation

ayende committed:
RavenDB-16446 Really fixing compilation now

arekpalinski commented on src/Raven.Server/Indexing/TempFileCache.cs:
```diff
@@ -0,0 +1,156 @@
+            string path = _options.TempPath.FullPath;
+            if (Directory.Exists(path) == false)
+                Directory.CreateDirectory(path);
+            foreach (string file in Directory.GetDirectories(path,  "lucene-*" + StorageEnvironmentOptions.DirectoryStorageEnvironmentOptions.TempFileExtension))
```
There shouldn't be any such temp files at this point because we delete them on storage startup. See https://github.com/ravendb/ravendb/pull/11497/files (https://issues.hibernatingrhinos.com/issue/RavenDB-15911)

ayende replied:
Yes, we don't expect that. However, I'm including this here in case the server crashed.
If that happened, we don't want the old files to still remain there (nothing will remove them).
For that purpose, we scan and remove / reuse them as needed.

ayende committed:
RavenDB-16446 Making sure that we open the files with DeleteOnClose if recovering existing temp files

ppekrol committed:
RavenDB-16446
- making sure that we are copying from the beginning of the MS to the FS
- making sure that the position of FS after copying the MS stays the same
- wrapping the FS to hide the real FS length

ayende committed:
RavenDB-16446 Fix computation of end of file on write after seek

ayende commented:
Probably need to merge that to 4.2 LTS as well
